Prometheus endpoint

seaninryan opened the pull request:
I've put together a new servlet to try and expose some of the existing stats (i.e. what's exposed by `http://<host>:3030/$/stats/<dataset>`) in a format consumable by prometheus.

With this change we'd see the metrics at `http://<host>:3030/$/prometheus` and they would look like this:
```
fuseki_requests_bad{operation="GSP_RW", description="Graph Store Protocol", dataset="/Argon", endpoint="data"} 0
fuseki_requests{operation="GSP_RW", description="Graph Store Protocol", dataset="/Argon", endpoint="data"} 1315
fuseki_requests_good{operation="GSP_RW", description="Graph Store Protocol", dataset="/Argon", endpoint="data"} 1315
fuseki_requests_bad{operation="Quads_RW", description="HTTP Quads", dataset="/Argon", endpoint=""} 0
fuseki_requests{operation="Quads_RW", description="HTTP Quads", dataset="/Argon", endpoint=""} 0
fuseki_requests_good{operation="Quads_RW", description="HTTP Quads", dataset="/Argon", endpoint=""} 0
fuseki_requests_bad{operation="Upload", description="File Upload", dataset="/Argon", endpoint="upload"} 0
fuseki_requests{operation="Upload", description="File Upload", dataset="/Argon", endpoint="upload"} 0
fuseki_requests_good{operation="Upload", description="File Upload", dataset="/Argon", endpoint="upload"} 0
fuseki_requests_bad{operation="Query", description="SPARQL Query", dataset="/Argon", endpoint="query"} 0
fuseki_requests{operation="Query", description="SPARQL Query", dataset="/Argon", endpoint="query"} 1694
fuseki_requests_good{operation="Query", description="SPARQL Query", dataset="/Argon", endpoint="query"} 1694
fuseki_requests_bad{operation="Query", description="SPARQL Query", dataset="/Argon", endpoint="sparql"} 0
...
etc
etc
```

I configured Prometheus to scrape every 5 seconds and in Grafana was able to put together these graphs

![2019-02-19-16-04-00_1836x379](https://user-images.githubusercontent.com/13253395/53029769-79b6b580-3461-11e9-84f8-e70428ae9088.png)

One the left is fuseki_requests grouped by dataset and on the right is the same metric but grouped by request type. I just wrote a curl loop of `GET`s to generate the spike

I on-purpose didn't introduce any new dependencies here - it'd be great to also expose jvm metrics like what happens in [the Spring Boot actuator for Prometheus](https://www.callicoder.com/spring-boot-actuator-metrics-monitoring-dashboard-prometheus-grafana/) but it would involve maybe adding dependencies on something like [micrometer](https://github.com/micrometer-metrics/micrometer)

I had looked into getting the tuple counts but I see that it's implemented via a sparql query in the Fuseki UI - so not sure how resource intensive it would be to continually do that

Let me know what you think - is there any interest out there in persevering with this?

kinow commented:
@Claudenw I think we had a discussion in the mailing list recently about Grafana, Prometheus. This might be interesting for you?

Claudenw commented:
@kinow thanks for the heads up.  @seaninryan works with me at the same company.  I advised him to make his changes known to the community for feedback and possible inclusion in next release.

ajs6f commented:
Is this endpoint usable by software other than Prometheus? I mean to ask; are there other monitoring suites that could accept this data or would we be catering especially to one system? I recognize that Prometheus is extremely high-profile and useful, just being cautious about playing any favorites…

seaninryan commented:
I'm not 100% sure @ajs6f  - but I think that it probably is a propriety Prometheus format. It's described [here](https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md) and that document doesn't make reference to any, say, standard etc.

Also spring actuator specifically calls its endpoint `/prometheus` and the [micrometer](https://github.com/micrometer-metrics/micrometer) library that spring boot depends on is called `micrometer-registry-prometheus`

ajs6f commented:
Ok, cool. Given the small amount of code here, I don't really have any problem with supporting one particular framework (especially since I understand Prometheus to be a very big player in that field). Just wanted to confirm. Thanks for the contribution @seaninryan !

afs commented:
micrometer.io looks interesting.

Have I understood micrometer correctly? The Fuseki counters/stats/... implement in the micrometer framework and then a choice of adapter makes that available to the specific package.

micrometer-core is about 0.5 Mbytes so not huge itself but it has dependencies (oddly, slf4j does not seem to isn't a dependency, log4j2 and logback are) - what might the size impact be?

Given the need to provide a choice of implementations, we have to have a way to allow choice but also work out of the box. Fuseki/full as a war file, and to some extent the server jar, are sealed so "some assembly required". Fuseki/main might be more tractable although again, the combined jar (the easy way to run it).

@ajs6f  - it used `org.pcollections` - is that like Dexx?

ajs6f commented:
It is like Dexx in that it is persistent (aka functional) collections library, and it is not like Dexx in that Dexx intends to exclusively port the Scala library types to Java, whereas PCollections is doing its own thing. I _think_ I took a look at PCollections when we were building up TIM, but I honestly can't remember.

Unlike Dexx, PCollections is clearly under active development. Might be time to take another look at it...

ajs6f commented:
Yep, [looks like](https://lists.apache.org/list.html?dev@jena.apache.org:gte=1d:pcollections) we did look at PCollections, and the performance wasn't too hot. But that was some time ago. Perhaps it's time to look again, or maybe look at some of the other alternatives under discussion early in TIM's creation. Ok, enough spamming this PR about data structures.

I wonder if we could split the difference with metrics by implementing Micrometer and allowing people to "bring their own" adapter in a JAR?

afs commented:
Yes - we have to allow "bring their own" adapter, and may provide one specific one (our own simple custom one?).

Is there any work around doing GraphQL for metrics?

seaninryan commented:
I changed what I'd done to use micrometer.. the creation of the required objects is a bit tricky so for now I've just done it in the servlet `init()` method

I just made it use the `JvmMemoryMetrics` class to pull in some standard stuff and added a `FusekiRequestsMetrics` class to get those stats that come from the `DataAccessPointRegistry`


I'd spoken to @Claudenw and looked into having a core dependency on `micrometer-core` and maybe an `extras` dependency on `micrometer-registry-prometheus` but I can't see how we'd instantiate the correct implementation of `MeterRegistry` because the different implementations all have constructors with different signatures

Anyway you guys might be able to offer some more insight on that... I've just been looking at spring-boot actuator and trying to see what they've done there

This is what the metrics look like with what I've set up so far:
```
# HELP jvm_buffer_count_buffers An estimate of the number of buffers in the pool
# TYPE jvm_buffer_count_buffers gauge
jvm_buffer_count_buffers{id="direct",} 16.0
jvm_buffer_count_buffers{id="mapped",} 51.0
# HELP jvm_memory_committed_bytes The amount of memory in bytes that is committed for the Java virtual machine to use
# TYPE jvm_memory_committed_bytes gauge
jvm_memory_committed_bytes{area="heap",id="PS Survivor Space",} 4.8758784E7
jvm_memory_committed_bytes{area="heap",id="PS Old Gen",} 2.8311552E8
jvm_memory_committed_bytes{area="heap",id="PS Eden Space",} 9.5944704E7
jvm_memory_committed_bytes{area="nonheap",id="Metaspace",} 3.2899072E7
jvm_memory_committed_bytes{area="nonheap",id="Code Cache",} 1.376256E7
jvm_memory_committed_bytes{area="nonheap",id="Compressed Class Space",} 4063232.0
# HELP jvm_buffer_total_capacity_bytes An estimate of the total capacity of the buffers in this pool
# TYPE jvm_buffer_total_capacity_bytes gauge
jvm_buffer_total_capacity_bytes{id="direct",} 81543.0
jvm_buffer_total_capacity_bytes{id="mapped",} 4.02663383E8
# HELP fuseki_requests_good  
# TYPE fuseki_requests_good gauge
fuseki_requests_good{dataset="/data-set-1",description="Graph Store Protocol",endpoint="data",operation="GSP_RW",} 154.0
fuseki_requests_good{dataset="/data-set-2",description="File Upload",endpoint="upload",operation="Upload",} 0.0
fuseki_requests_good{dataset="/data-set-2",description="Graph Store Protocol (Read)",endpoint="get",operation="GSP_R",} 0.0
fuseki_requests_good{dataset="/data-set-1",description="HTTP Quads",endpoint="",operation="Quads_RW",} 0.0
fuseki_requests_good{dataset="/data-set-2",description="HTTP Quads",endpoint="",operation="Quads_RW",} 0.0
fuseki_requests_good{dataset="/data-set-2",description="SPARQL Query",endpoint="sparql",operation="Query",} 0.0
fuseki_requests_good{dataset="/data-set-1",description="File Upload",endpoint="upload",operation="Upload",} 0.0
fuseki_requests_good{dataset="/data-set-1",description="SPARQL Update",endpoint="update",operation="Update",} 2.0
fuseki_requests_good{dataset="/data-set-1",description="Graph Store Protocol (Read)",endpoint="get",operation="GSP_R",} 0.0
fuseki_requests_good{dataset="/data-set-2",description="Graph Store Protocol",endpoint="data",operation="GSP_RW",} 0.0
fuseki_requests_good{dataset="/data-set-2",description="SPARQL Query",endpoint="query",operation="Query",} 0.0
fuseki_requests_good{dataset="/data-set-2",description="SPARQL Update",endpoint="update",operation="Update",} 0.0
fuseki_requests_good{dataset="/data-set-1",description="SPARQL Query",endpoint="query",operation="Query",} 178.0
fuseki_requests_good{dataset="/data-set-1",description="SPARQL Query",endpoint="sparql",operation="Query",} 0.0
# HELP jvm_buffer_memory_used_bytes An estimate of the memory that the Java virtual machine is using for this buffer pool
# TYPE jvm_buffer_memory_used_bytes gauge
jvm_buffer_memory_used_bytes{id="direct",} 81543.0
jvm_buffer_memory_used_bytes{id="mapped",} 4.02663383E8
# HELP fuseki_requests  
# TYPE fuseki_requests gauge
fuseki_requests{dataset="/data-set-1",description="Graph Store Protocol",endpoint="data",operation="GSP_RW",} 154.0
fuseki_requests{dataset="/data-set-2",description="File Upload",endpoint="upload",operation="Upload",} 0.0
fuseki_requests{dataset="/data-set-2",description="Graph Store Protocol (Read)",endpoint="get",operation="GSP_R",} 0.0
fuseki_requests{dataset="/data-set-1",description="HTTP Quads",endpoint="",operation="Quads_RW",} 0.0
fuseki_requests{dataset="/data-set-2",description="HTTP Quads",endpoint="",operation="Quads_RW",} 0.0
fuseki_requests{dataset="/data-set-2",description="SPARQL Query",endpoint="sparql",operation="Query",} 0.0
fuseki_requests{dataset="/data-set-1",description="File Upload",endpoint="upload",operation="Upload",} 0.0
fuseki_requests{dataset="/data-set-1",description="SPARQL Update",endpoint="update",operation="Update",} 2.0
fuseki_requests{dataset="/data-set-1",description="Graph Store Protocol (Read)",endpoint="get",operation="GSP_R",} 0.0
fuseki_requests{dataset="/data-set-2",description="Graph Store Protocol",endpoint="data",operation="GSP_RW",} 0.0
fuseki_requests{dataset="/data-set-2",description="SPARQL Query",endpoint="query",operation="Query",} 0.0
fuseki_requests{dataset="/data-set-2",description="SPARQL Update",endpoint="update",operation="Update",} 0.0
fuseki_requests{dataset="/data-set-1",description="SPARQL Query",endpoint="query",operation="Query",} 178.0
fuseki_requests{dataset="/data-set-1",description="SPARQL Query",endpoint="sparql",operation="Query",} 0.0
# HELP jvm_memory_max_bytes The maximum amount of memory in bytes that can be used for memory management
# TYPE jvm_memory_max_bytes gauge
jvm_memory_max_bytes{area="heap",id="PS Survivor Space",} 4.8758784E7
jvm_memory_max_bytes{area="heap",id="PS Old Gen",} 1.431830528E9
jvm_memory_max_bytes{area="heap",id="PS Eden Space",} 6.16562688E8
jvm_memory_max_bytes{area="nonheap",id="Metaspace",} -1.0
jvm_memory_max_bytes{area="nonheap",id="Code Cache",} 2.5165824E8
jvm_memory_max_bytes{area="nonheap",id="Compressed Class Space",} 1.073741824E9
# HELP fuseki_requests_bad  
# TYPE fuseki_requests_bad gauge
fuseki_requests_bad{dataset="/data-set-1",description="Graph Store Protocol",endpoint="data",operation="GSP_RW",} 0.0
fuseki_requests_bad{dataset="/data-set-2",description="File Upload",endpoint="upload",operation="Upload",} 0.0
fuseki_requests_bad{dataset="/data-set-2",description="Graph Store Protocol (Read)",endpoint="get",operation="GSP_R",} 0.0
fuseki_requests_bad{dataset="/data-set-1",description="HTTP Quads",endpoint="",operation="Quads_RW",} 0.0
fuseki_requests_bad{dataset="/data-set-2",description="HTTP Quads",endpoint="",operation="Quads_RW",} 0.0
fuseki_requests_bad{dataset="/data-set-2",description="SPARQL Query",endpoint="sparql",operation="Query",} 0.0
fuseki_requests_bad{dataset="/data-set-1",description="File Upload",endpoint="upload",operation="Upload",} 0.0
fuseki_requests_bad{dataset="/data-set-1",description="SPARQL Update",endpoint="update",operation="Update",} 0.0
fuseki_requests_bad{dataset="/data-set-1",description="Graph Store Protocol (Read)",endpoint="get",operation="GSP_R",} 0.0
fuseki_requests_bad{dataset="/data-set-2",description="Graph Store Protocol",endpoint="data",operation="GSP_RW",} 0.0
fuseki_requests_bad{dataset="/data-set-2",description="SPARQL Query",endpoint="query",operation="Query",} 0.0
fuseki_requests_bad{dataset="/data-set-2",description="SPARQL Update",endpoint="update",operation="Update",} 0.0
fuseki_requests_bad{dataset="/data-set-1",description="SPARQL Query",endpoint="query",operation="Query",} 0.0
fuseki_requests_bad{dataset="/data-set-1",description="SPARQL Query",endpoint="sparql",operation="Query",} 0.0
# HELP jvm_memory_used_bytes The amount of used memory
# TYPE jvm_memory_used_bytes gauge
jvm_memory_used_bytes{area="heap",id="PS Survivor Space",} 2.1266496E7
jvm_memory_used_bytes{area="heap",id="PS Old Gen",} 8.859756E7
jvm_memory_used_bytes{area="heap",id="PS Eden Space",} 1.3267632E7
jvm_memory_used_bytes{area="nonheap",id="Metaspace",} 3.1906112E7
jvm_memory_used_bytes{area="nonheap",id="Code Cache",} 1.3602752E7
jvm_memory_used_bytes{area="nonheap",id="Compressed Class Space",} 3776040.0
```

I need to look into how/whether what I've done would automatically pick up a new dataset and things like that

The great thing about the subclasses of `MeterBinder` is that there are loads of existing dashboards on grafana that can be used to chart the metrics. [This one](https://grafana.com/dashboards/6756) for example. Also loads of them are applicable to fuseki
![2019-02-25-19-46-17_480x475](https://user-images.githubusercontent.com/13253395/53364154-18e21e00-3936-11e9-808e-59ab7552e980.png)

seaninryan commented on jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/ctl/ActionPrometheus.java:
```diff
@@ -0,0 +1,68 @@
+            action.response.setContentType( contentTypeJSON );
+            action.response.setCharacterEncoding( charsetUTF8 );
+
+            out.write( meterRegistry.scrape().getBytes() );
```
this scrape method is only on the Prometheus implementation of `MeterRegistry` :confused:

seaninryan commented:
Hi all

I think I've got this to somewhere decent now
- I've got a core dependency on `micrometer-core` and have add a module in extras which pulls in `micrometer-registry-prometheus`
  - so the extras module created a shaded jar that you can just drop in
- I've created an extras servlet that maps to `$/extras/*`
  - this picks up on classes that implement `ExtraAction` which gives a path that gets mapped on startup. So `ActionPrometheus` (in the extras module) returns the string `prometheus` which means we have a mapping of `$/extras/prometheus`
  - I wrote it to pick up the exact path - so we might want to extend to, say make a url particular to a dataset or allow wildcards of its own?
  - have a look at the `ActionExtras` class... it kind of breaks with the conventions of other similar classes by not calling `doCommon()`... I'm not sure if that's a big deal but I think `/$/extras` is incompatible with requiring a dataset name etc. You guys will know better I guess
- I just picked a bunch of easy to set up `MeterBinders` that seemed relevant. There are a couple of Jetty ones that I'd like to add but they require access to some Jetty objects (haven't looked too deeply) that I don't have access to right now

Anyway you can get really nice dashboards [like this](https://grafana.com/dashboards/4701) for free then
![2019-02-27-16-52-21_1889x956](https://user-images.githubusercontent.com/13253395/53507629-23282780-3ab0-11e9-8ab5-be932c95c8cd.png)

I've just zoomed out to give an overall effect... these graphs are just memory usage, cpu etc.

It would be great if you guys could give this the once over tell me what I need to do to get it considered for merging to master

afs commented:
This is looking very interesting!

I'll look at the servlet when I'm at my development machine (not this coming week). There are already freestanding, non-dataset servlets with a Fuseki lifecycle which adds access to the logging, action tracking, and also counters (which is a bit recursive!). It isn't necessary to use this framework - plain servlets work - though if it can be, it is nice to collect everything together.

Fuseki/Full is a webapp war file and runs in Tomcat as well as with the Jetty packaging so Jetty-specific features need to adapt to Jetty not being used. Fuseki/main, the adaptable programmable version, no UI at the moment, is a standalone server that uses Jetty and isn't a webapp war file deployment. These two systems are about packaging and delivery of the common core which is the bulk of Fuseki.

seaninryan commented:
Thanks for looking @afs :+1:

I'm not 100% clear on how all the packaging works. Be great if you can fire it up next week and let me know what you think

afs commented:
This is looking good.  I have it working in Fuseki Full (webapp form) and in Fuseki Main (the more configurable servlet form) - details for running in Fuseki Main below.

#### Configuration

Something we have to think about carefully is using `ServiceLoader`. Jena already has an [init mechanism](https://jena.apache.org/documentation/notes/system-initialization.html) and it allows controlling the order of initialization. It is quite simple - there is an `init()` method which might be a limitation.

`ActionPrometheus` could use this to wire in a function lambda to call it's `perform()` into a registry.

I don't know at the moment which of a "different ServiceLoader" or "single ServiceLoader" is better.

#### Naming

Rather than hide under `/$/extras/` I think I prefer `/$/metrics/` or something else to cover just statistics reporting. This is a significant functionality, not to be hidden under "extras".

#### Dispatch

The `ActionExtras.perform` could dispatch on the final component of the URI.  No fixed string!

It might even be better to look for the dispatch string in `init(ServletConfig)` to find the prefix for this servlet instance.

#### HTTP MIME Type

The HTTP return is set as `application/json;charset=utf-8` but it isn't JSON. If there isn't a specific MIME type, then `text/plain;charset=utf-8` would be better (?).

#### jena-prometheus jar

The shaded JAR has got slf4j classes in it. I don't think they are needed. Can they be excluded
? either in the POM or in the shader. Maybe it is an effect of the `<scope>provided</scope>`.

[jena-fuseki2/jena-fuseki-server/pom.xml](https://github.com/apache/jena/blob/master/jena-fuseki2/jena-fuseki-server/pom.xml) has an example of shading Jena including license control.

#### Licensing

`org.hdrhistogram` and  `org.latencyutils` are 2-clause BSD license and we a shipping them inshared artifacts so the acknowledge clause applies.   No problem, just a task to do.

#### Fuseki Main example

Illustration of running in `FusekiMain`:

    public static void mainServer(String ... a) {
        FusekiLogging.setLogging();
        FusekiServer server = FusekiServer.create()
            .add("/ds", DatasetGraphFactory.createTxnMem())
            .port(3030)
            .verbose(true)
            .addServlet("/$/extras/*", new ActionExtras())
            .build();
        
        try { server.start().join(); }
        finally { server.stop(); }
    }

Claudenw commented:
*bump* 

Does anyone have in their plans to look at this?

afs commented:
#### Packaging 

Looking in more detail ... the duplication of classes in the jena-prometheus jar means the amount of new code in module `jena-promotheus` is quite small (it is local code and prometheus-registry).  By my rough calculation, 

In Fuseki core:
```
416K        io/micrometer/micrometer-core/
112K        org/hdrhistogram/
32K         org/latencyutils/
```
In jena-prometheus:
```
28K         io/micrometer/micrometer-registry-prometheus/
12K         jena-prometheus jar
```
(have I got that right?)

Having to repackage Fuseki-full to get a WAR file or a new combined jar is an extra step for the end user. This was the step that took me the longest to get something working.

The code modularity is worth it - the packaging less so IMO.

afs commented:
Admin: [JENA-1691](https://issues.apache.org/jira/browse/JENA-1691).

seaninryan commented:
@afs I've pushed a bunch of changes based off your feedback. I think it's quite neat now apart from the creation of the _combined jar_ as you've mentioned above.

I got rid of `/$/extras/*` and just went with `/$/metrics` and by default that will return `501 Not Implemented` because the default metrics provider doesn't provide any equivalent of the `scrape()` method that Prometheus provides

Anyway, take a look and let me know what you think

seaninryan commented:
@afs - w.r.t. licenses for `HdrHistogram` and `LatencyUtils` - what's involved there? Downloading and copying in the `LICENCE` files from their github projects or something else?

afs commented:
#### Licensing

@claudew - would you be able to help out with the L&N issues here?

The LICENSE and NOTICE for the combined binary artifacts (3 of them) wil need to include BSD-related text.
At the moment, they get autogenerated but we need to put specific text in.

jena-fuseki-server
jena-fuseki-war
jena-fuseki-fulljar

See jena-fuseki-webapp/src/resources/META-INF/ which puts in LICENSE and NOTICE into a jar but I think we need to do it in the shading step for external resources like BSD-licnesed binaries. 

#### Packaging

It is easier to add maven modules later than retire them.
Prometheus is sufficiently common I think including in Fuseki core for now is easier for users. 

Proposal: moving it to
jena-fuseki-core::org.apache.jena.fuseki.metrics.prometheus (and the ServiceLoader file similarly)
and now have the jena-prometheus (which is out of sequence in the build currently).

#### Testing

If `jena-prometheus` is not available I get a `500 server error` and a stacktrace, not 501 (running Fuseki main+Prometheus)

{noformat}
[12:47:13] INFO  Server               :: Start Fuseki (port=3030)
[12:47:18] WARN  HttpChannel          :: /$/metrics/ds
org.apache.jena.fuseki.servlets.ActionErrorException: SimpleMeterRegistry isn't scrapeable
	at org.apache.jena.fuseki.servlets.ServletOps.error(ServletOps.java:143)
	at org.apache.jena.fuseki.servlets.ServletOps.errorNotImplemented(ServletOps.java:116)
	at org.apache.jena.fuseki.metrics.SimpleMetricsProvider.scrape(SimpleMetricsProvider.java:36)
	at org.apache.jena.fuseki.ctl.ActionMetrics.perform(ActionMetrics.java:51)
	at org.apache.jena.fuseki.ctl.ActionMetrics.doGet(ActionMetrics.java:42)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
{noformat}

Some tests of the metrics code would be good.  Testing the server is tricky because you have to have a server and `jena-fuseki-core` isn't a whole server. test can go into jena-fuseki-main for functional testing and ideally  jena-fuseki-webapp (just to test web.xml/shiro.ini setup).

#### Code comment

The direct adding of `MeterRegistry` to everywhere `DataAccessPointRegistry` is created feels wrong :-) but this does not have to be looked at now. The current way works.

claudew commented:
@afs I think you meant @claudenw (Claude Warren), not @claudew (Claude Whitmyer).

afs commented:
@claudew - yes, sorry. 

@Claudenw - would you be able to help out with the L&N issues here?

seaninryan commented:
:+1: thanks @afs I've removed the new extras module.

I made the Prometheus provider the default for now - I think we may as well seeing as Prometheus stuff is included in `jena-fuseki-core` now. Does that make sense?

I'll have a look at the 501 error thing later - or maybe can delete the `SimpleMetricsProvider` given that it can't be scraped via http (not sure when anyone would use it for real t.b.h)

... and the testing. Hope to get to it later or tomorrow

seaninryan commented:
Sorted the 501 thing... will look at the tests tomorrow

seaninryan commented:
I've added a very basic test @afs

afs commented on jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/metrics/SimpleMetricsProvider.java:
```diff
@@ -33,7 +34,8 @@ public MeterRegistry getMeterRegistry() {
     public void scrape(HttpAction action) {
-        ServletOps.errorNotImplemented( "SimpleMeterRegistry isn't scrapeable" );
+        ServletOps.responseSendError(action.response, HttpSC.NOT_IMPLEMENTED_501,
+                "SimpleMeterRegistry isn't scrapeable" );
```
Something strange is going on.
`ServletOps.errorNotImplemented` should do that via throwing `ActionErrorException` which is caught in `ActionBase`. Ah - `ActionMetrics.doGet` is doing it's own thing not using the common lifecycle.

Using `ServletOps.errorNotImplemented` and the `ActionMetrics` below for the common execution c. Is there anything in the lifecycle that was a problem?

```
public class ActionMetrics extends ActionCtl {
    
    public ActionMetrics() { }

    @Override
    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
        super.doCommon(req, resp);
    }

    @Override
    protected void perform(HttpAction action) {
        MetricsProviderRegistry.get().scrape( action );
    }
}
```

seaninryan replied:
Yeah that works great - I've just pushed the change.

The previous implementation was left over from when I was doing that `/$/extras/*` stuff that's gone now

afs commented:
One test goes a long way!

afs commented on jena-fuseki2/jena-fuseki-webapp/src/test/java/org/apache/jena/fuseki/TestMetrics.java:
```diff
@@ -0,0 +1,68 @@
+/**
```
Add TestMetrics.class to \@Suite.SuiteClasses in TS_FusekiWebapp.

seaninryan replied:
Changed :+1:

afs commented on jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/metrics/prometheus/PrometheusMetricsProvider.java:
```diff
@@ -0,0 +1,77 @@
+        try (ServletOutputStream out = action.response.getOutputStream()) {
+            action.response.setContentType( WebContent.contentTypeTextPlain );
+            action.response.setCharacterEncoding( WebContent.charsetUTF8 );
+
```
Add `ServletOps.success(action);` just before the writing starts. This will record the response code. Otherwise "-1 -1" gets printed to the Fuseki log.

seaninryan replied:
Changed :+1:

Claudenw commented:
I think we only need to add the licenses (there may be only one) to the
META-INF folder.  We could do this by just copying the file(s) or we could
extract them from the Jars duirng the build.

The first is easier, the latter means that if there are any changes in the
license files we will pick them up.

Is there any desire to do this one way or the other?  If not I will
probably just copy the files.

The other issue may be the license scan (at least I think we do one) will
have to be updated to accept the licenses.

Does this only go into the fuseki war or are there other builds that need
them too?

Claude

On Thu, Apr 4, 2019 at 3:32 PM Andy Seaborne <notifications@github.com>
wrote:

> Licensing
>
> @claudew <https://github.com/claudew> - would you be able to help out
> with the L&N issues here?
>
> The LICENSE and NOTICE for the combined binary artifacts (3 of them) wil
> need to include BSD-related text.
> At the moment, they get autogenerated but we need to put specific text in.
>
> jena-fuseki-server
> jena-fuseki-war
> jena-fuseki-fulljar
>
> See jena-fuseki-webapp/src/resources/META-INF/ which puts in LICENSE and
> NOTICE into a jar but I think we need to do it in the shading step for
> external resources like BSD-licnesed binaries.
> Packaging
>
> It is easier to add maven modules later than retire them.
> Prometheus is sufficiently common I think including in Fuseki core for now
> is easier for users.
>
> Proposal: moving it to
> jena-fuseki-core::org.apache.jena.fuseki.metrics.prometheus (and the
> ServiceLoader file similarly)
> and now have the jena-prometheus (which is out of sequence in the build
> currently).
> Testing
>
> If jena-prometheus is not available I get a 500 server error and a
> stacktrace, not 501 (running Fuseki main+Prometheus)
>
> {noformat}
> [12:47:13] INFO Server :: Start Fuseki (port=3030)
> [12:47:18] WARN HttpChannel :: /$/metrics/ds
> org.apache.jena.fuseki.servlets.ActionErrorException: SimpleMeterRegistry
> isn't scrapeable
> at org.apache.jena.fuseki.servlets.ServletOps.error(ServletOps.java:143)
> at
> org.apache.jena.fuseki.servlets.ServletOps.errorNotImplemented(ServletOps.java:116)
> at
> org.apache.jena.fuseki.metrics.SimpleMetricsProvider.scrape(SimpleMetricsProvider.java:36)
> at org.apache.jena.fuseki.ctl.ActionMetrics.perform(ActionMetrics.java:51)
> at org.apache.jena.fuseki.ctl.ActionMetrics.doGet(ActionMetrics.java:42)
> at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
> {noformat}
>
> Some tests of the metrics code would be good. Testing the server is tricky
> because you have to have a server and jena-fuseki-core isn't a whole
> server. test can go into jena-fuseki-main for functional testing and
> ideally jena-fuseki-webapp (just to test web.xml/shiro.ini setup).
> Code comment
>
> The direct adding of MeterRegistry to everywhere DataAccessPointRegistry
> is created feels wrong :-) but this does not have to be looked at now. The
> current way works.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/jena/pull/534#issuecomment-479922355>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACU6nuyFp_lZM4a13rHdSMlU3oKi6tNfks5vdg0SgaJpZM4bDW9i>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

afs commented:
> Does this only go into the fuseki war 

@Claudenw There are three binary artifacts.

Claudenw commented:
I was goingto add the licensing information but discovered that the
HdrHistogram and LatencyUtils are both licensed under creative commons
public domain (http://creativecommons.org/publicdomain/zero/1.0/) if we
don't want to use that they also provide the BSD 2-Clause License.  I
suggest we use creative commons public domain because it sounds easier.  In
which case I don't think we need to do anything.

Claude




On Thu, Apr 4, 2019 at 3:32 PM Andy Seaborne <notifications@github.com>
wrote:

> Licensing
>
> @claudew <https://github.com/claudew> - would you be able to help out
> with the L&N issues here?
>
> The LICENSE and NOTICE for the combined binary artifacts (3 of them) wil
> need to include BSD-related text.
> At the moment, they get autogenerated but we need to put specific text in.
>
> jena-fuseki-server
> jena-fuseki-war
> jena-fuseki-fulljar
>
> See jena-fuseki-webapp/src/resources/META-INF/ which puts in LICENSE and
> NOTICE into a jar but I think we need to do it in the shading step for
> external resources like BSD-licnesed binaries.
> Packaging
>
> It is easier to add maven modules later than retire them.
> Prometheus is sufficiently common I think including in Fuseki core for now
> is easier for users.
>
> Proposal: moving it to
> jena-fuseki-core::org.apache.jena.fuseki.metrics.prometheus (and the
> ServiceLoader file similarly)
> and now have the jena-prometheus (which is out of sequence in the build
> currently).
> Testing
>
> If jena-prometheus is not available I get a 500 server error and a
> stacktrace, not 501 (running Fuseki main+Prometheus)
>
> {noformat}
> [12:47:13] INFO Server :: Start Fuseki (port=3030)
> [12:47:18] WARN HttpChannel :: /$/metrics/ds
> org.apache.jena.fuseki.servlets.ActionErrorException: SimpleMeterRegistry
> isn't scrapeable
> at org.apache.jena.fuseki.servlets.ServletOps.error(ServletOps.java:143)
> at
> org.apache.jena.fuseki.servlets.ServletOps.errorNotImplemented(ServletOps.java:116)
> at
> org.apache.jena.fuseki.metrics.SimpleMetricsProvider.scrape(SimpleMetricsProvider.java:36)
> at org.apache.jena.fuseki.ctl.ActionMetrics.perform(ActionMetrics.java:51)
> at org.apache.jena.fuseki.ctl.ActionMetrics.doGet(ActionMetrics.java:42)
> at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
> {noformat}
>
> Some tests of the metrics code would be good. Testing the server is tricky
> because you have to have a server and jena-fuseki-core isn't a whole
> server. test can go into jena-fuseki-main for functional testing and
> ideally jena-fuseki-webapp (just to test web.xml/shiro.ini setup).
> Code comment
>
> The direct adding of MeterRegistry to everywhere DataAccessPointRegistry
> is created feels wrong :-) but this does not have to be looked at now. The
> current way works.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/jena/pull/534#issuecomment-479922355>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACU6nuyFp_lZM4a13rHdSMlU3oKi6tNfks5vdg0SgaJpZM4bDW9i>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

Claudenw commented:
I can not find a straight text version of
https://creativecommons.org/publicdomain/zero/1.0/legalcode so perhaps we
should just provide a link to it at the bottom of the current license like
we do with the other included code.

On Sat, Apr 6, 2019 at 4:59 PM Claude Warren <claude@xenei.com> wrote:

> I was goingto add the licensing information but discovered that the
> HdrHistogram and LatencyUtils are both licensed under creative commons
> public domain (http://creativecommons.org/publicdomain/zero/1.0/) if we
> don't want to use that they also provide the BSD 2-Clause License.  I
> suggest we use creative commons public domain because it sounds easier.  In
> which case I don't think we need to do anything.
>
> Claude
>
>
>
>
> On Thu, Apr 4, 2019 at 3:32 PM Andy Seaborne <notifications@github.com>
> wrote:
>
>> Licensing
>>
>> @claudew <https://github.com/claudew> - would you be able to help out
>> with the L&N issues here?
>>
>> The LICENSE and NOTICE for the combined binary artifacts (3 of them) wil
>> need to include BSD-related text.
>> At the moment, they get autogenerated but we need to put specific text in.
>>
>> jena-fuseki-server
>> jena-fuseki-war
>> jena-fuseki-fulljar
>>
>> See jena-fuseki-webapp/src/resources/META-INF/ which puts in LICENSE and
>> NOTICE into a jar but I think we need to do it in the shading step for
>> external resources like BSD-licnesed binaries.
>> Packaging
>>
>> It is easier to add maven modules later than retire them.
>> Prometheus is sufficiently common I think including in Fuseki core for
>> now is easier for users.
>>
>> Proposal: moving it to
>> jena-fuseki-core::org.apache.jena.fuseki.metrics.prometheus (and the
>> ServiceLoader file similarly)
>> and now have the jena-prometheus (which is out of sequence in the build
>> currently).
>> Testing
>>
>> If jena-prometheus is not available I get a 500 server error and a
>> stacktrace, not 501 (running Fuseki main+Prometheus)
>>
>> {noformat}
>> [12:47:13] INFO Server :: Start Fuseki (port=3030)
>> [12:47:18] WARN HttpChannel :: /$/metrics/ds
>> org.apache.jena.fuseki.servlets.ActionErrorException: SimpleMeterRegistry
>> isn't scrapeable
>> at org.apache.jena.fuseki.servlets.ServletOps.error(ServletOps.java:143)
>> at
>> org.apache.jena.fuseki.servlets.ServletOps.errorNotImplemented(ServletOps.java:116)
>> at
>> org.apache.jena.fuseki.metrics.SimpleMetricsProvider.scrape(SimpleMetricsProvider.java:36)
>> at org.apache.jena.fuseki.ctl.ActionMetrics.perform(ActionMetrics.java:51)
>> at org.apache.jena.fuseki.ctl.ActionMetrics.doGet(ActionMetrics.java:42)
>> at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
>> {noformat}
>>
>> Some tests of the metrics code would be good. Testing the server is
>> tricky because you have to have a server and jena-fuseki-core isn't a
>> whole server. test can go into jena-fuseki-main for functional testing and
>> ideally jena-fuseki-webapp (just to test web.xml/shiro.ini setup).
>> Code comment
>>
>> The direct adding of MeterRegistry to everywhere DataAccessPointRegistry
>> is created feels wrong :-) but this does not have to be looked at now. The
>> current way works.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/apache/jena/pull/534#issuecomment-479922355>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/ACU6nuyFp_lZM4a13rHdSMlU3oKi6tNfks5vdg0SgaJpZM4bDW9i>
>> .
>>
>
>
> --
> I like: Like Like - The likeliest place on the web
> <http://like-like.xenei.com>
> LinkedIn: http://www.linkedin.com/in/claudewarren
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

afs commented:
A link in LICENSE is better and acceptable these days. HdrHistogram and LatencyUtils binaries artifacts themselves don't include a LICENSE or NOTICE file (these would automatically get rolled up by the shader). 

This is the LICENSE to end up in the convenience binaries, not the Jena one; nothing to do there - no "other license" source in the jena source codebase. And because it is the convenience binaries, and licenses with no attribute requirement, do we need to do anything?

Claudenw commented:
After reviewing the licensing documentation in the libraries and Apache's
policy on public domain licenses (
http://www.apache.org/legal/resolved.html#handling-public-domain-licensed-works)
I think we can safely include them as we would any other Apache approved
license.  To my mind this means adding a line or two to the Notice file
indicating that the two components are included and are licensed under the
createive commons public domain license.  This may be a bit of overkill,
but better safe than sorry.

Claude

On Sun, Apr 7, 2019 at 4:53 PM Andy Seaborne <notifications@github.com>
wrote:

> A link in LICENSE is better and acceptable these days. HdrHistogram and
> LatencyUtils binaries artifacts themselves don't include a LICENSE or
> NOTICE file (these would automatically get rolled up by the shader).
>
> This is the LICENSE to end up in the convenience binaries, not the Jena
> one; nothing to do there - no "other license" source in the jena source
> codebase. And because it is the convenience binaries, and licenses with no
> attribute requirement, do we need to do anything?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/jena/pull/534#issuecomment-480602750>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACU6nsUwPa171J5_JmaqNPn8of-U3Dgpks5vehRvgaJpZM4bDW9i>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

afs commented:
It is a ASF policy that NOTICE is minimal. This is to avoid simply passing the problems onto downstream consumers, because NOTICE is legal obligations and downstream would then have to handle it (recusively). LatenacyUtils and HdrHistogram don't have NOTICE files in the source repo.

http://www.apache.org/dev/licensing-howto.html#step-by-step
"Add any _mandatory_ legal notifications specific to the IP of your product to NOTICE." (my emphasis).

If there is no attribution requirement, NOTICE should not be touched. BSD has such a requirement, CC0 does not.

It sounds like the appropriate LICENSE is needed.
@Claudenw - do you agree?

(That the artifacts themselves don't have a LICENSE is a nuisance but it looks like we ought to put one into our combined binaries.)

Related - no dates! Avoid anything that needs maintenance!

afs commented:
See [JENA-1703](https://issues.apache.org/jira/browse/JENA-1703), PR #555 for nearby changes in notice files.

Claudenw commented:
Andy, thanks for getting the changes in, I agree with your assessment,
sorry that it took so long to respond.  So the Prometheus endpoint will
need to add 2 more licence notes to the files you created in PR #555
<https://github.com/apache/jena/pull/555>.  Is that your understanding as
well?

Claude

On Sat, Apr 13, 2019 at 2:20 PM Andy Seaborne <notifications@github.com>
wrote:

> See JENA-1703 <https://issues.apache.org/jira/browse/JENA-1703>, PR #555
> <https://github.com/apache/jena/pull/555> for nearby changes in notice
> files.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/jena/pull/534#issuecomment-482808571>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACU6npVhe8nwvnnzxkZfVGbehk_6IAJJks5vgdmGgaJpZM4bDW9i>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

afs commented:
From the legal-discuss@ emailing list, the decision seems to be we are using 2 clause BSD. Please confirm that here or dev@ so it is in public, on a jena list/forum.

Then we need to execute on that for the two dependencies that will be bundled into shared jars because we unpack and repack them.

Claudenw commented:
The two modules in question come under dual licenses: CC0 or
BSD-2-clause.  My understanding from the legal-discuss thread is that
the CC0 does not require notification because it is public domain but
it also does not confer patent rights, so we elected to use
BSD-2-Clause and need to add the BSD-2-Clause license file and a the
notation that it applies to the  HdrHistogram and LatencyUtils
packages.

I assume the notation will be in the Notice files for the fuseki
bundles that include the libraries mentioned..

Claude

On Sun, Apr 14, 2019 at 1:56 PM Andy Seaborne <notifications@github.com> wrote:
>
> From the legal-discuss@ emailing list, the decision seems to be we are using 2 clause BSD. Please confirm that here or dev@ so it is in public, on a jena list/forum.
>
> Then we need to execute on that for the two dependencies that will be bundled into shared jars because we unpack and repack them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.



-- 
I like: Like Like - The likeliest place on the web
LinkedIn: http://www.linkedin.com/in/claudewarren

Claudenw commented:
What is the process for me (3rd party) to add licenses to PR?

On Mon, Apr 15, 2019 at 7:35 AM Claude Warren <claude@xenei.com> wrote:
>
> The two modules in question come under dual licenses: CC0 or
> BSD-2-clause.  My understanding from the legal-discuss thread is that
> the CC0 does not require notification because it is public domain but
> it also does not confer patent rights, so we elected to use
> BSD-2-Clause and need to add the BSD-2-Clause license file and a the
> notation that it applies to the  HdrHistogram and LatencyUtils
> packages.
>
> I assume the notation will be in the Notice files for the fuseki
> bundles that include the libraries mentioned..
>
> Claude
>
> On Sun, Apr 14, 2019 at 1:56 PM Andy Seaborne <notifications@github.com> wrote:
> >
> > From the legal-discuss@ emailing list, the decision seems to be we are using 2 clause BSD. Please confirm that here or dev@ so it is in public, on a jena list/forum.
> >
> > Then we need to execute on that for the two dependencies that will be bundled into shared jars because we unpack and repack them.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
>
>
>
> --
> I like: Like Like - The likeliest place on the web
> LinkedIn: http://www.linkedin.com/in/claudewarren



-- 
I like: Like Like - The likeliest place on the web
LinkedIn: http://www.linkedin.com/in/claudewarren

afs commented:
@Claudenw Would it be possible for you to work with Sean to get it in the PR?

Claudenw commented:
I hope to do that in the next few days / evenings


On Wed, Apr 17, 2019 at 3:59 PM Andy Seaborne <notifications@github.com>
wrote:

> @Claudenw <https://github.com/Claudenw> Would it be possible for you to
> work with Sean to get it in the PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/jena/pull/534#issuecomment-484127012>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACU6nnppkUMY6uOy25Q17y6ZYs5zS6lCks5vhza6gaJpZM4bDW9i>
> .
>


-- 
I like: Like Like - The likeliest place on the web
<http://like-like.xenei.com>
LinkedIn: http://www.linkedin.com/in/claudewarren

ajs6f commented:
@Claudenw It can be a little tricky, but you can send a PR to @seaninryan's PR. You would do that by pulling his branch into your local Github repo, and starting a new branch from that. Add the appropriate commits, then send a a PR via Github, selecting his branch as the base branch and your new branch as the comparison branch. He can then accept your PR into his branch, which will automatically incorporate it here, where that branch is being used for a PR to Jena.

That might or might not be easier for you, just throwing it out there in case it's useful.

seaninryan commented:
Hi @afs, @claude - I've added licensing info to the `jena-fuseki2/LICENSE` file

Is this enough? I notice that the top level `NOTICE`  contains copyright info for some dependencies. Should I copy the HdrHistogram and LatencyUtils copyrights into that file? Or maybe add a `NOTICE` to the `jena-fuseki2` module?

afs commented:
@seaninryan - the material needs to go in the NOTICE file (in merged PR #554).

But the information is there so I'll merge this and fix up the 3 NOTICE files so we can get a release done with this in.

seaninryan commented:
ok great @afs thanks for all of your time

afs commented:
Subsequent NOTICE updates done with commit e1a392ef3035 .

@seaninryan - thank you!

afs commented:
Documentation: this does not need to be sync'ed to a release cycle. We publish the website on release but can (and do) publish at other times to get stuff out.

ajs6f commented:
Just want to second @afs-- thank you @seaninryan, metrics are becoming more and more important and it's great to have this for Jena!
